soroban-env-host/src/storage.rs

```diff
@@ -7,17 +7,24 @@
 //!   - [Env::put_contract_data](crate::Env::put_contract_data)
 //!   - [Env::del_contract_data](crate::Env::del_contract_data)
 
+use std::cmp::Ordering;
 use std::rc::Rc;
 
-use soroban_env_common::Compare;
+use soroban_env_common::{Compare, RawVal};
 
 use crate::budget::Budget;
-use crate::xdr::{LedgerEntry, LedgerKey, ScHostStorageErrorCode};
+use crate::xdr::{LedgerEntry, LedgerKey, ScHostStorageErrorCode, ScVal};
 use crate::Host;
 use crate::{host::metered_map::MeteredOrdMap, HostError};
 
 pub type FootprintMap = MeteredOrdMap<Rc<LedgerKey>, AccessType, Budget>;
 pub type StorageMap = MeteredOrdMap<Rc<LedgerKey>, Option<Rc<LedgerEntry>>, Budget>;
+// Alias for the `RawVal` that has the cheap 'shallow' comparison defined - it
+// just compares the payloads without looking at the value semantics.
+// This is used in `TempStorageMap` to make sure that we don't do expensive
+// value comparisons.
+pub type ShallowComparableRawVal = RawVal;
```

Comment:
This doesn't define a different type (for the sake of impls, for example). The `impl<ShallowComparableRawVal> for Budget` below is impl'ed on `RawVal` too, the two types are considered identical by this line.

You probably want `struct ShallowComparableRawVal(RawVal)` or .. something. But I am also not sure why you're doing this. Map-key comparison is actually supposed to be deep, otherwise you can put an entry under, say, `vec [1,2]` and later fail to retrieve it by looking it up with `vec[1,2]` with a superficially different object-handle identity.

Reply:
This is not for keys, this is for values. I'm actually thinking that we probably need to come back to https://github.com/stellar/rs-soroban-env/pull/624 - I still believe that the value comparison requirement doesn't make sense for the map because the values should never be compared (unless we have a bug and allow non-unique entries). Instead it should do sorting only by keys.

Reply:
As I said over there: all scvals are comparable because they can occur as keys in maps, because maps are polymorphic -- they can be keyed by any scval. You can make a map that is keyed by other maps. It's not _common_ but we'd have to develop much more elaborate special-casing of comparable and non-comparable scvals to prohibit it. the general case is "all scvals are comparable".

Reply:
This is not about scvals or any types being comparable. The issue is that currently the map requires comparison for *values* (not just keys). I got convinced that it's ok-ish in the initial discussion in #624, but now that I think about it, it seems just wrong because values should never be compared in the map (even if we implement map via sorted tuple vector) and if they are, then something is wrong with the implementation (unless the map is a multimap, that is).

Reply:
Ok. I think you shouldn't have to impl anything here at all -- `Compare<RawVal>` is impl'ed already -- but regardless I've redone (and expanded a bit) the task you set on in #624 in #739 and am ok with it if you really feel strongly.

Reply:
Yes, I've switched to `RawVal`s for keys and removed this alias completely.